salt/modules/file.py

```diff
@@ -2676,56 +2691,72 @@ def _add_content(linesep, lines=None, include_marker_start=True,
             block_found = True
         elif append_if_not_found:
             # Make sure we have a newline at the end of the file
-            if 0 != len(new_file):
+            if new_file:
                 if not new_file[-1].endswith(linesep):
                     new_file[-1] += linesep
             # add the markers and content at the end of file
             _add_content(linesep, lines=new_file)
             block_found = True
-        else:
-            raise CommandExecutionError(
-                'Cannot edit marked block. Markers were not found in file.'
-            )
-
-    if block_found:
-        diff = ''.join(difflib.unified_diff(orig_file, new_file))
-        has_changes = diff is not ''
-        if has_changes and not dry_run:
-            # changes detected
-            # backup file attrs
-            perms = {}
-            perms['user'] = get_user(path)
-            perms['group'] = get_group(path)
-            perms['mode'] = salt.utils.files.normalize_mode(get_mode(path))
-
-            # backup old content
-            if backup is not False:
-                backup_path = '{0}{1}'.format(path, backup)
-                shutil.copy2(path, backup_path)
-                # copy2 does not preserve ownership
-                check_perms(backup_path,
-                        None,
-                        perms['user'],
-                        perms['group'],
-                        perms['mode'])
+        elif insert_before_match or insert_after_match:
+            match_regex = insert_before_match or insert_after_match
+            if not isinstance(match_regex, six.string_types):
+                raise CommandExecutionError(
+                    'RegEx expected in match parameter.'
+                )
```

Comment:
Why wait until this late to check for a regex? This should be done much earlier, when we are checking the arguments passed to the function. Additionally, the error message doesn't make sense as there is no `match` parameter. We should be telling the user _which_ parameter is invalid, instead of making them figure it out.